### PR TITLE
DiscoveryCommunity

### PR DESCRIPTION
**Pull request by egbertbouman**

### CODE REVIEW

File: discovery/community.py [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+
+    def __str__(self):
+        return "PTB_%d_%d_%s_%s" % (self.timestamp, self.overlap, self.candidate_mid.encode("HEX"), self.received_from)
+
+    def __hash__(self):
+        return hash(self.candidate_mid)
+
+class DiscoveryCommunity(Community):
+
+    def initialize(self, attach=True, max_prefs=25, max_tbs=25):
```
Comment (NielsZeilemaker) [inline discovery/community.py]: The attach keyword is removed, you can remove it and hence don't pass it to the super

File: discovery/community.py [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+            # even when success is False it is still possible that *some* addresses were resolved
+            for sock_addr in self.bootstrap.candidates:
+                print >> sys.stderr, "Adding", sock_addr, "as discovered candidate"
+                self.add_discovered_candidate(Candidate(sock_addr, False))
+
+            if success:
+                logger.debug("resolved all bootstrap addresses")
+
+        bootstrap_file = os.path.join(self._dispersy._working_directory, "bootstraptribler.txt")
+        print >> sys.stderr, "Expecting bootstrapfile at", os.path.abspath(bootstrap_file), os.path.exists(bootstrap_file)
```
Comment (NielsZeilemaker) [inline discovery/community.py]: Could you remove this print, or convert it to a logger.debug

**Comment by NielsZeilemaker** [comment]
In addition to that remark, I guess we need to upgrade all print >> sys.stderr's to logger.debug calls.

File: discovery/conversion.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+from random import choice, sample
+from binascii import hexlify, unhexlify
+from socket import inet_ntoa, inet_aton
+
+from ..message import DropPacket
+from ..conversion import BinaryConversion
+from ..bloomfilter import BloomFilter
+
+
+def bytes_to_long(val, nrbytes=0):
```
Comment (NielsZeilemaker) [inline discovery/conversion.py]: I'm guessing this isn't used anymore

File: discovery/test.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+from Tribler.dispersy.community import Community
+from Tribler.community.privatesemantic.community import DiscoveryCommunity
+
+
+class NoFSemanticCommunity(DiscoveryCommunity, Community):
```
Comment (NielsZeilemaker) [inline discovery/test.py]: This file can go

File: tests/dispersytestclass.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 from twisted.internet.defer import inlineCallbacks, returnValue
 from twisted.internet.threads import blockingCallFromThread
 
-from ..bootstrap import Bootstrap
+from ..discovery.bootstrap import Bootstrap
```
Comment (NielsZeilemaker) [inline tests/dispersytestclass.py]: Instead of setting Bootstrap to False/True, we should pass autoload_discover=False/True to dispersy.start()

**Comment by whirm** [comment]
@egbertbouman could you wrap the lines you modified to 120char max?

**Comment by whirm** [comment]
Other than that I didn't see anything obvious.

**Comment by whirm** [comment]
if @NielsZeilemaker is happy with it, we can merge already and Ill rebase my work

File: tests/dispersytestclass.py [hunk 6]
```diff
@@ -53,16 +50,16 @@ def tearDown(self):
             logger.warning("Failing")
         assert not pending, "The reactor was not clean after shutting down all dispersy instances."
 
-    def create_nodes(self, amount=1, store_identity=True, tunnel=False, communityclass=DebugCommunity):
+    def create_nodes(self, amount=1, store_identity=True, tunnel=False, communityclass=DebugCommunity, autoload_discovery=True):
```
Comment (NielsZeilemaker) [inline tests/dispersytestclass.py]: Could you change this one to autoload=False

File: tests/test_bootstrap.py [hunk 7]
```diff
@@ -69,15 +68,15 @@ def dispersy_enable_candidate_walker_responses(self):
             node, = self.create_nodes(1, communityclass=Community)
```
Comment (NielsZeilemaker) [inline tests/test_bootstrap.py]: autoload=True

File: tests/dispersytestclass.py (hunk 5 above)
Comment (NielsZeilemaker) [inline tests/dispersytestclass.py]: Can you remove the import

File: tests/test_bootstrap.py [hunk 8]
```diff
@@ -113,8 +112,8 @@ def __init__(self, *args, **kargs):
                 self._summary = {}
                 self._hostname = {}
                 self._identifiers = {}
-                self._pcandidates = self._dispersy._bootstrap_candidates.values()
```
Comment (NielsZeilemaker) [inline tests/test_bootstrap.py:116]: This isn't ok, we need to use the bootstrap candidates as they are being used in the discoverycommunity

File: tests/dispersytestclass.py [hunk 9]
```diff
@@ -33,7 +28,8 @@ def setUp(self):
 
         self.assertFalse(reactor.getDelayedCalls())
         self._mm = None
-        self._mm, = self.create_nodes()
+        # Don't load DiscoveryCommunity so it doesn't mess with the tests
```
Comment (NielsZeilemaker) [inline tests/dispersytestclass.py]: autoload_discovery is False by default, could you revert these lines changes

File: tests/test_bootstrap.py (hunk 8 above)
Comment (NielsZeilemaker) [inline tests/test_bootstrap.py:116]: I guess you need to call self._dispersy.get_communities() and then check with isinstance to get the discoverycommunity.

**Pull request merged**